pkg/types/ccipocr3/generic_types.go

```diff
@@ -82,51 +82,85 @@ func (c ChainSelector) String() string {
-	Data []byte `json:"data"`
-}
-
-type TokenAmount struct {
-	Token  types.Account
-	Amount *big.Int
-}
-
-func (c CCIPMsg) String() string {
+// Message is the generic Any2Any message type for CCIP messages.
+// It represents, in particular, a message emitted by a CCIP onramp.
+// The message is expected to be consumed by a CCIP offramp after
+// translating it into the appropriate format for the destination chain.
+type Message struct {
+	// Header is the family-agnostic header for OnRamp and OffRamp messages.
+	// This is always set on all CCIP messages.
+	Header RampMessageHeader `json:"header"`
+
+	// EVM2AnyRampMessage is populated if the CCIP message originated from an EVM chain.
+	EVM2AnyRampMessage *EVM2AnyRampMessage `json:"evm2AnyRampMessage,omitempty"`
+}
+
+func (c Message) String() string {
 	js, _ := json.Marshal(c)
 	return string(js)
 }
 
-type CCIPMsgBaseDetails struct {
-	// ID is a unique identifier for the message, it should be unique across all chains.
+// RampMessageHeader is the family-agnostic header for OnRamp and OffRamp messages.
+// The MessageID is not expected to match MsgHash, since it may originate from a different
+// ramp family.
+type RampMessageHeader struct {
+	// MessageID is a unique identifier for the message, it should be unique across all chains.
 	// It is generated on the chain that the CCIP send is requested (i.e. the source chain of a message).
-	ID string `json:"id"`
-	// SourceChain is the chain that the message originated from.
-	SourceChain ChainSelector `json:"sourceChain,string"`
-	// SeqNum is an auto-incrementing sequence number for the message.
-	// NOTE: Sequence numbers are unique per chain. Meaning that the same sequence number can exist on multiple chains.
-	SeqNum SeqNum `json:"seqNum,string"`
+	MessageID Bytes32 `json:"messageId"`
+	// SourceChainSelector is the chain selector of the chain that the message originated from.
+	SourceChainSelector ChainSelector `json:"sourceChainSelector,string"`
+	// DestChainSelector is the chain selector of the chain that the message is destined for.
+	DestChainSelector ChainSelector `json:"destChainSelector,string"`
+	// SequenceNumber is an auto-incrementing sequence number for the message.
+	// Not unique across lanes.
+	SequenceNumber SeqNum `json:"seqNum,string"`
+	// Nonce is the nonce for this lane for this sender, not unique across senders/lanes
+	Nonce uint64 `json:"nonce"`
 
 	// MsgHash is the hash of all the message fields.
 	// NOTE: The field is expected to be empty, and will be populated by the plugin using the MsgHasher interface.
 	MsgHash Bytes32 `json:"msgHash"` // populated
+
+	// OnRamp is the address of the onramp that sent the message.
+	// NOTE: This is populated by the ccip reader. Not emitted explicitly onchain.
+	OnRamp Bytes `json:"onRamp"`
+}
+
+// EVM2AnyRampMessage is the family-agnostic message sent from an EVM onramp
+// to all other chain families.
+type EVM2AnyRampMessage struct {
```

Comment:
Some thoughts on type-per-family approach.

1. You're using bytes for everything, thats already universal so why bother with custom types? I'd recommend moving all of these into the header and either (a) putting chain specific fields in `Data` or having an `Extensions` field for chain specific fields to be added in later, `ExtraArgs` could also serve this purpose.

2. This approach would lead to a lot of nearly duplicate code because you'd have to select the type and then perform the same business logic. Are you thinking that we'll have some parallel type that swaps in some sort of multi-addr type?

3. I'm not sure this even makes sense. The Sender and Receiver addresses are exactly (one of) the problems we need to solve for  different families. They're both included in the same data structure so why bother splitting out a type like this?

4. IMO this data structure should contain at most the fields specific to the destination, in other words `Any2EVM` rather than `EVM2Any`.

Reply:
Updated based on your feedback and offline discussion.

Reply:
Changes look good. The only nitpick I have left is that maybe some of these other fields could go to the header. i.e. Sender/Receiver and maybe fee info.

Reply:
I was keeping Header this way to try and remain consistent with the onchain header, but not sure if there's much value in that (we're already divergent given the fact that we have the message hash and the onramp address in the header, which is not in the onchain header)

Reply:
Fee info should probably not go to the header, since it's an OnRamp detail (not needed for OffRamp)

For the rest I think it might be easier to stay consistent w/ OnChain? We do not include sender / receiver in the header because we use specific types OnChain (i.e. one of them is not always bytes - `address sender` for `EVM2Any` and `address receiver` for `Any2EVM`)

Reply:
Yeah I guess the `Bytes` type for both sender and receiver here are pretty opaque compared w/ onchain. I think its fine to keep it this way if only for consistency and maintaining the same mental model.

Reply:
I pretty much agree. There was a comment `// This is always set on all CCIP messages.` which had me thinking that these are the fields which are always set.